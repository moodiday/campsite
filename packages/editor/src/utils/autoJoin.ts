// https://discuss.prosemirror.net/t/how-to-autojoin-all-the-time/2957/4

import { NodeType } from '@tiptap/pm/model'
import { Transaction } from '@tiptap/pm/state'
import { canJoin } from '@tiptap/pm/transform'

// Ripped out from prosemirror-commands wrapDispatchForJoin
export function autoJoin({
  prevTr,
  nextTr,
  nodeTypes
}: {
  prevTr: Transaction
  nextTr: Transaction
  nodeTypes: NodeType[]
}) {
  // Find all ranges where we might want to join.
  let ranges: Array<number> = []

  for (let i = 0; i < prevTr.mapping.maps.length; i++) {
    let map = prevTr.mapping.maps[i]

    if (!map) continue
    for (let j = 0; j < ranges.length; j++) ranges[j] = map.map(ranges[j]!)
    map.forEach((_s, _e, from, to) => ranges.push(from, to))
  }

  // Figure out which joinable points exist inside those ranges,
  // by checking all node boundaries in their parent nodes.
  let joinable: number[] = []

  for (let i = 0; i < ranges.length; i += 2) {
    let from = ranges[i],
      to = ranges[i + 1]
    let $from = prevTr.doc.resolve(from!),
      depth = $from.sharedDepth(to!),
      parent = $from.node(depth)

    for (let index = $from.indexAfter(depth), pos = $from.after(depth + 1); pos <= to!; ++index) {
      let after = parent.maybeChild(index)

      if (!after) {
        break
      }
      if (index && joinable.indexOf(pos) == -1) {
        let before = parent.child(index - 1)

        if (before.type == after.type && nodeTypes.includes(before.type)) {
          joinable.push(pos as number)
        }
      }
      pos += after.nodeSize
    }
  }

  let joined = false

  // Join the joinable points
  joinable.sort((a, b) => a - b)
  for (let i = joinable.length - 1; i >= 0; i--) {
    if (canJoin(prevTr.doc, joinable[i]!)) {
      nextTr.join(joinable[i]!)
      joined = true
    }
  }

  return joined
}
