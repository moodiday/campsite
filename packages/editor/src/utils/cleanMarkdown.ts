export default function cleanMarkdown(text: string): string {
  const re = /^\s?(\[(X|\s|_|-)\]\s(.*)?)/gim

  while (text.match(re)) {
    text = text.replace(re, (match) => `- ${match.trim()}`)
  }

  return text.replace(/\n{3,}/g, '\n\n\\\n').replace(/\b\n\b/g, '\n\n')
}
