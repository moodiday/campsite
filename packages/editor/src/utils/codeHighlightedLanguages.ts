import { refractor } from 'refractor'
import bash from 'refractor/lang/bash.js'
import clike from 'refractor/lang/clike.js'
import clojure from 'refractor/lang/clojure.js'
import cpp from 'refractor/lang/cpp.js'
import csharp from 'refractor/lang/csharp.js'
import css from 'refractor/lang/css.js'
import elixir from 'refractor/lang/elixir.js'
import erlang from 'refractor/lang/erlang.js'
import go from 'refractor/lang/go.js'
import graphql from 'refractor/lang/graphql.js'
import groovy from 'refractor/lang/groovy.js'
import haskell from 'refractor/lang/haskell.js'
import hcl from 'refractor/lang/hcl.js'
import ini from 'refractor/lang/ini.js'
import java from 'refractor/lang/java.js'
import javascript from 'refractor/lang/javascript.js'
import json from 'refractor/lang/json.js'
import jsx from 'refractor/lang/jsx.js'
import kotlin from 'refractor/lang/kotlin.js'
import lisp from 'refractor/lang/lisp.js'
import lua from 'refractor/lang/lua.js'
import markup from 'refractor/lang/markup.js'
import nix from 'refractor/lang/nix.js'
import objectivec from 'refractor/lang/objectivec.js'
import ocaml from 'refractor/lang/ocaml.js'
import perl from 'refractor/lang/perl.js'
import php from 'refractor/lang/php.js'
import powershell from 'refractor/lang/powershell.js'
import python from 'refractor/lang/python.js'
import ruby from 'refractor/lang/ruby.js'
import rust from 'refractor/lang/rust.js'
import sass from 'refractor/lang/sass.js'
import scala from 'refractor/lang/scala.js'
import scss from 'refractor/lang/scss.js'
import solidity from 'refractor/lang/solidity.js'
import sql from 'refractor/lang/sql.js'
import swift from 'refractor/lang/swift.js'
import toml from 'refractor/lang/toml.js'
import tsx from 'refractor/lang/tsx.js'
import typescript from 'refractor/lang/typescript.js'
import verilog from 'refractor/lang/verilog.js'
import vhdl from 'refractor/lang/vhdl.js'
import visualbasic from 'refractor/lang/visual-basic.js'
import yaml from 'refractor/lang/yaml.js'
import zig from 'refractor/lang/zig.js'

export const ALIAS_TO_LANGUAGE = [
  [bash, 'Bash'] as const,
  [clojure, 'Clojure'] as const,
  [cpp, 'C++'] as const,
  [css, 'CSS'] as const,
  [clike, 'C-Like'] as const,
  [csharp, 'C#'] as const,
  [elixir, 'Elixir'] as const,
  [erlang, 'Erlang'] as const,
  [go, 'Go'] as const,
  [graphql, 'GraphQL'] as const,
  [groovy, 'Groovy'] as const,
  [haskell, 'Haskell'] as const,
  [hcl, 'HCL'] as const,
  [ini, 'INI'] as const,
  [java, 'Java'] as const,
  [javascript, 'JavaScript'] as const,
  [jsx, 'JSX'] as const,
  [json, 'JSON'] as const,
  [kotlin, 'Kotlin'] as const,
  [lisp, 'Lisp'] as const,
  [lua, 'Lua'] as const,
  [markup, 'HTML'] as const,
  [nix, 'Nix'] as const,
  [objectivec, 'Objective-C'] as const,
  [ocaml, 'OCaml'] as const,
  [perl, 'Perl'] as const,
  [php, 'PHP'] as const,
  [python, 'Python'] as const,
  [powershell, 'PowerShell'] as const,
  [ruby, 'Ruby'] as const,
  [rust, 'Rust'] as const,
  [scala, 'Scala'] as const,
  [sql, 'SQL'] as const,
  [solidity, 'Solidity'] as const,
  [sass, 'Sass'] as const,
  [scss, 'SCSS'] as const,
  [swift, 'Swift'] as const,
  [toml, 'TOML'] as const,
  [typescript, 'TypeScript'] as const,
  [tsx, 'TSX'] as const,
  [verilog, 'Verilog'] as const,
  [vhdl, 'VHDL'] as const,
  [visualbasic, 'Visual Basic'] as const,
  [yaml, 'YAML'] as const,
  [zig, 'Zig'] as const
].reduce(
  (acc, [language, name]) => {
    refractor.register(language)
    acc[language.displayName] = name
    language.aliases.forEach((alias) => {
      acc[alias] = name
    })
    return acc
  },
  { none: 'Plain Text' } as Record<string, string>
)
