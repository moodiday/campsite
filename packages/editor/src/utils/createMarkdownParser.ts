import { AnyExtension } from '@tiptap/core'
import { Schema } from '@tiptap/pm/model'
import MarkdownIt from 'markdown-it'
import Token from 'markdown-it/lib/token'

import markdownItTasks from './markdownItTasks'
import { MarkdownParser, ParseSpec } from './MarkdownParser'

const ignoreAllMarkdownTokens: { [name: string]: ParseSpec } = {
  // blocks
  paragraph: { block: 'paragraph' },
  blockquote: { block: 'paragraph' },
  code_block: { block: 'paragraph' },
  fence: { block: 'paragraph' },
  heading: { block: 'paragraph' },
  hr: { ignore: true, noCloseToken: true },
  ordered_list: { ignore: true },
  bullet_list: { ignore: true },
  list_item: { ignore: true },
  references: { block: 'paragraph' },
  table: { block: 'paragraph' },
  thead: { ignore: true },
  th: { ignore: true },
  tbody: { ignore: true },
  tr: { ignore: true },
  td: { ignore: true },
  image: { block: 'paragraph', noCloseToken: true },
  hardbreak: { ignore: true, noCloseToken: true },
  softbreak: { ignore: true, noCloseToken: true },

  // inline
  code_inline: { ignore: true },
  strong: { ignore: true },
  em: { ignore: true },
  link: { ignore: true },
  s: { ignore: true }
}

export function listIsTight(tokens: readonly Token[], i: number) {
  while (++i < tokens.length) if (tokens[i].type != 'list_item_open') return tokens[i].hidden
  return false
}

function getMarkdownSpec(extension: AnyExtension) {
  if ('markdownParseSpec' in extension.config === false || typeof extension.config.markdownParseSpec !== 'function') {
    return null
  }

  return extension.config.markdownParseSpec() as ParseSpec | null | undefined
}

function getMarkdownToken(extension: AnyExtension) {
  if ('markdownToken' in extension.config === false || typeof extension.config.markdownToken !== 'string') {
    return extension.name
  }
  return extension.config.markdownToken
}

export const createMarkdownParserSpec = (spec: ParseSpec) => spec

export function createMarkdownParser(
  schema: Schema,
  extensions: AnyExtension[],
  // WARNING: Provide JSDOM window and document objects to avoid errors in Node.js environment
  domParser: DOMParser = new window.DOMParser(),
  document: Document = window.document
) {
  const tokens = extensions
    .filter(
      (extension) => 'markdownParseSpec' in extension.config && typeof extension.config.markdownParseSpec === 'function'
    )
    .reduce((nodes, extension) => {
      const parseSpec = getMarkdownSpec(extension)

      if (!parseSpec) {
        return nodes
      }

      return {
        ...nodes,
        [getMarkdownToken(extension)]: parseSpec
      }
    }, ignoreAllMarkdownTokens)

  return new MarkdownParser(
    schema,
    MarkdownIt('default', {
      // enables raw inline html with the html_inline token
      html: true,
      breaks: false,
      linkify: true
    }).use(markdownItTasks),
    tokens,
    domParser,
    document
  )
}
