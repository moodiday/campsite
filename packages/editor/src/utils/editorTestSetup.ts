export function editorTestSetup() {
  function getBoundingClientRect(): DOMRect {
    const rec = {
      x: 0,
      y: 0,
      bottom: 0,
      height: 0,
      left: 0,
      right: 0,
      top: 0,
      width: 0
    }

    return { ...rec, toJSON: () => rec }
  }

  class FakeDOMRectList extends Array<DOMRect> implements DOMRectList {
    item(index: number): DOMRect | null {
      return this[index]
    }
  }

  document.elementFromPoint = (): null => null
  HTMLElement.prototype.getBoundingClientRect = getBoundingClientRect
  HTMLElement.prototype.getClientRects = (): DOMRectList => new FakeDOMRectList()
  Range.prototype.getBoundingClientRect = getBoundingClientRect
  Range.prototype.getClientRects = (): DOMRectList => new FakeDOMRectList()
}
