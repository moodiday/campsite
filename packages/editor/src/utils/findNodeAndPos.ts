import { Node as ProseMirrorNode } from '@tiptap/pm/model'
import { EditorState } from '@tiptap/pm/state'

export function findNodeAndPos(state: EditorState, match: (node: ProseMirrorNode) => boolean) {
  let result: { pos: number; node: ProseMirrorNode } | undefined

  state.doc.descendants((node, pos) => {
    if (match(node)) {
      result = { pos, node }
      return false
    }
    return true
  })
  return result
}
