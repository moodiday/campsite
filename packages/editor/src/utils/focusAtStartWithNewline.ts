import { Editor } from '@tiptap/core'

export function focusAtStartWithNewline(editor: Editor) {
  if (editor.state.doc.firstChild?.type.name !== 'paragraph' || editor.state.doc.firstChild?.textContent !== '') {
    const tr = editor.state.tr.insert(0, editor.schema.nodes.paragraph.create())

    editor.view.dispatch(tr)
  }
  editor.commands.focus('start', { scrollIntoView: true })
}
