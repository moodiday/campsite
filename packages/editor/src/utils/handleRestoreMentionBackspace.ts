import { EditorState, Transaction } from '@tiptap/pm/state'

export function handleRestoreMentionBackspace({
  transaction,
  state,
  nodeName,
  char
}: {
  transaction: Transaction
  state: EditorState
  nodeName: string
  char: string
}) {
  let isMention = false
  const { selection } = state
  const { empty, anchor } = selection

  if (!empty) {
    return false
  }

  state.doc.nodesBetween(anchor - 1, anchor, (node, pos) => {
    if (node.type.name === nodeName) {
      isMention = true
      transaction.insertText(char, pos, pos + node.nodeSize)

      return false
    }
  })

  return isMention
}
