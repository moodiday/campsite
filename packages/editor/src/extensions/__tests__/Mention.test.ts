import { Editor } from '@tiptap/core'
import { describe, expect, it } from 'vitest'

import { editorTestSetup } from '../../utils/editorTestSetup'
import * as E from '../index'

editorTestSetup()

function setupEditor(content: string) {
  return new Editor({
    extensions: [E.Document, E.Text, E.Paragraph, E.Mention],
    content
  })
}

describe('Mention', () => {
  it('inserts before the paragraph', () => {
    const editor = setupEditor('<p>Foo bar</p>')
    // mimic TipTap Mention command
    // https://github.com/ueberdosis/tiptap/blob/main/packages/extension-mention/src/mention.ts#L48-L61

    editor
      .chain()
      .focus()
      .insertContent([
        {
          type: E.Mention.name,
          attrs: { id: 'm-id', label: 'm-label', username: 'm-username' }
        },
        { type: 'text', text: ' ' }
      ])
      .run()

    // match against HTML as our API depends on the HTML structure not changing
    expect(editor.getHTML()).toMatchSnapshot()
  })
})
