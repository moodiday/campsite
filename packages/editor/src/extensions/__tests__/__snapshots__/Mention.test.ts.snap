// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mention > inserts before the paragraph 1`] = `"<p><span class="mention" data-type="mention" data-id="m-id" data-label="m-label" data-role="member" data-username="m-username">@m-label</span> Foo bar</p>"`;
